Ad kernels

keniley1 opened the pull request:
This is a rather large PR, but it's pretty simple. It adds some basic AD kernels to the repository, along with an input file using AD with Zapdos and Crane together in the problems directory as an example. The Crane submodule is also updated. 

**Added AD Materials:**
- ADGasElectronMoments

**Added AD Kernels:**
- ADCoeffDiffusion
- ADEFieldAdvection
- ADJouleHeating
- ADTimeDerivativeLog

**Added AD BCs:**
- ADDCIonBC
- ADHagelaarElectronBC
- ADHagelaarEnergyBC
- ADHagelaarIonAdvectionBC
- ADHagelaarIonDiffusionBC
- ADNeumannCircuitVoltageMoles_KV
- ADSecondaryElectronBC
- ADSecondaryElectronEnergyBC

**Crane Submodule Update**
- Added functionality to have multiple action blocks added (one for water and one for gas, for example)
- Added AD functionality with the "use_ad = true" input parameter. At the moment this only applies to Townsend coefficients and their associated kernels (This requires using ADGasElectronMoments in Zapdos)

keniley1 commented:
Unfortunately, the Crane update required a pretty big refactoring of the Action. The good news is that it will be infinitely easier to update in the future if anything is changed; everything is now wrapped in functions so it's much easier to read. From a user standpoint this is irrelevant, but it's easier for me to debug if something goes wrong!

The bad news is that this required renaming some Crane materials and kernels. @csdechant this means that your input files are outdated. I updated all of the tests you included so they are working again. Sorry about any inconvenience. If you aren't already,

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/484735/) on 9c36380 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/zapdos/docs/PRs/54/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/zapdos/docs/PRs/54/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 90f8f2aea641680de6b3da9614cca762a7008d68`

keniley1 commented:
The PR failed the clang formatting precheck, which is odd...I made sure to run clang-formatting on every single file in this directory before I submitted. I'm not sure what's failing. The suggested patch isn't working.

cticenhour commented:
@keniley1 what is the output of `which clang-format` on your machine?

keniley1 commented:
/usr/bin/clang-format

cticenhour commented:
And you are using Clang/LLVM as your compiler? I have it as default on my mac, and if I remember right, git integration for clang-format comes with the LLVM compiler.

cticenhour commented:
Also it seems as though you're getting that curl error because the patch doesn't exist on the server now.

@milljm Any reason why wouldn't copy a clang-format patch or just delete it? I'm getting a "Not Found" error if I try to go to the link directly.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/484898/) on 5ced272 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/zapdos/docs/PRs/54/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/zapdos/docs/PRs/54/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 90f8f2aea641680de6b3da9614cca762a7008d68`

keniley1 commented:
@cticenhour Alright, for some reason my Ubuntu workstation was not properly linking to all of the moose dependencies...I may need to do a fresh moose install. Weird. I've never had any issues until now. 
Anyway, I manually added the /opt/moose/llvm... to my path and the git clang-format worked. 

Unfortunately I'm now _still_ failing the precheck, and the git clang-format command returns "clang-format did not modify any files". I'm lost.

milljm commented:
@cticenhour because no one has created a ZAPDOS alias in the apache configuration file on mooseframework.org

Makes me wonder if there are any other apps that needs adjusting. Give me a minute to add this project.

cticenhour commented:
@keniley1 That is weird. Might be best to do an update. Eventually we're switching MOOSE environment distribution to use conda, but some bugs are still being worked out of that approach. Hopefully when that happens environment issues can be more easily fixed without going through the pain of a full re-install. 

Anyway, you're not the first to have weirdness happen with clang-format. I wonder if it has something to do with the version of clang-format on CIVET vs the version on your machine. Do you mind if I try to apply a patch to this PR?

keniley1 commented:
@cticenhour Maybe not necessary - I tracked down the single offending file and fixed it myself. We'll see if the tests work!

milljm commented:
@keniley1 @cticenhour Okay... I've updated the Apache config. Those links should work.

cticenhour commented:
@keniley1 Awesome - testing is going now. I was glancing through the changed files and had a question - did you put a test in for the new AD Kernels? I see actions tests, but wasn't sure if these new kernels were included in those.

keniley1 commented:
@cticenhour Whoops - no, there's no test for those yet. I absolutely meant to add one. I already have a working input file with AD in the problems directory so I'll simplify that and throw it into a test. 

Unfortunately exodiff doesn't like my input files. I've been getting rid of the domain scaling and using automatic scaling instead. This works perfectly, but now since some of my elements are O(10nm) exodiff throws an error. I guess that is below the resolution of the test. Let's see if I can fix that...

Thanks @milljm !

keniley1 commented:
@csdechant Honestly, I have no idea how that happened. That Kernel was renamed "EEDFElasticLog" but all I did was find and replace the names. I have no clue how those just got removed. I'll add them back in. Sorry about that!

cticenhour commented:
@keniley1 So with automatic scaling you haven't had to do *any* manual scaling (in terms of the domain scale) at all? That super nice!

keniley1 commented:
@cticenhour Yes, pretty much! I still use kV scaling because I'm used to it, but I usually run simulations without any spatial scaling whatsoever. 

My tests are failing here but run fine on my desktop and laptop. I'm not sure how to fix that.

keniley1 commented:
I checked this PR out on my Mac, and sure enough, my tests failed. I'm almost able to fix this by relaxing the tolerances to 1e-4 and fixing the floor tolerance to ~1e-8 or some small, nonzero value since some of the differences are between 0 and 1e-218...which yeah, I guess that's _technically_ a 100% (or infinite?) difference but come on... 

But some of the AuxVariables are still failing. This is very strange to me. In particular, for my townsend test (townsend_units.i) the only differences are with rho and DiffusiveFlux_em, and they seem significant. In the early timesteps the error in DiffusiveFlux_em is 200% because of a sign flip, e.g. -1.47e6 compared to 1.47e6. Later on it's usually the difference between O(10^4) and 0 (which seems huge, but keep in mind diffusive flux gets up to 10^18 by the end of the simulation; 10^4 is practically a rounding error at that point). I'm not sure what to do here. The actual variable values are identical so I have no idea why these AuxVariables are different, and I don't feel comfortable just chalking this up to an OS difference between Ubuntu 18 and Mac Catalina because these exact same AuxVariables are used in other tests and don't have this issue at all. 

Anyone have any ideas?

lindsayad commented:
What are your solve tolerances?

Have you considered making a .cmp file?

> On Mar 19, 2020, at 7:04 AM, Shane Keniley <notifications@github.com> wrote:
> 
> ﻿
> I checked this PR out on my Mac, and sure enough, my tests failed. I'm almost able to fix this by relaxing the tolerances to 1e-4 and fixing the floor tolerance to ~1e-8 or some small, nonzero value since some of the differences are between 0 and 1e-218...which yeah, I guess that's technically a 100% (or infinite?) difference but come on...
> 
> But some of the AuxVariables are still failing. This is very strange to me. In particular, for my townsend test (townsend_units.i) the only differences are with rho and DiffusiveFlux_em, and they seem significant. In the early timesteps the error in DiffusiveFlux_em is 200% because of a sign flip, e.g. -1.47e6 compared to 1.47e6. Later on it's usually the difference between O(10^4) and 0 (which seems huge, but keep in mind diffusive flux gets up to 10^18 by the end of the simulation; 10^4 is practically a rounding error at that point). I'm not sure what to do here. The actual variable values are identical so I have no idea why these AuxVariables are different, and I don't feel comfortable just chalking this up to an OS difference between Ubuntu 18 and Mac Catalina because these exact same AuxVariables are used in other tests and don't have this issue at all.
> 
> Anyone have any ideas?
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

keniley1 commented:
> What are your solver tolerances? 
For the problem input file the relative tolerance is 1e-5 with an absolute tolerance of 1e-4. In this case the domain is scaled as it is in mean_en.i (e.g. gas domain scaled by 1e-3, liquid domain by 1e-7). 

> Have you considered making a .cmp file?
Yes, when I said that I can fix it by relaxing the relative and floor tolerances, that's what I meant. This does not work with the auxiliary variables here though; I don't have a good relative or absolute tolerance to use across all AuxVariables. 

It's just very strange to me that only AuxVariables are different. I would expect the nonlinear variables to be different since these both depend strongly on Arp and em, but they're the same. I don't get it.

lindsayad commented:
>  I don't have a good relative or absolute tolerance to use across all AuxVariables.

With a custom cmp you should be able to set different tolerances for each variable

lindsayad commented:
> Edit: This issue also only exists during the transient. There appears to be no difference in the final steady state solution.

In this case you could just execute outputs on final

keniley1 commented:
Sorry for all the commits - the standards have been tightened since the last time I submitted a PR! I'll make the next one cleaner. 

> With a custom cmp you should be able to set different tolerances for each variable

I had no idea. I've always just run exodiff -summary, which by default groups nodal and elemental variables together with a single tolerance. I assumed that's just how it cmp files worked. Good to know!

cticenhour commented:
> Sorry for all the commits - the standards have been tightened since the last time I submitted a PR! I'll make the next one cleaner.

Yeah....the debug builds were added by me after your last PR I think, so that we wouldn't go back to the dark days of not being able to build in debug mode. 😄 

You could do a squash of commits if the commit number bothers you, but I don't mind it as-is.

keniley1 commented:
It really was bothering me a little bit. I squashed all of the ones related to tests and clang formatting!

lindsayad commented on problems/ad_argon_chemistry/reaction_em + Ar -> em + Ar*.txt:
```diff
@@ -0,0 +1,200 @@
+2.9270e-01 0.0000e+00 
```
Are these spaces in the file name? Gross!

keniley1 replied:
Yeah, it's ugly...I don't like it either. It was just for convenience at first and I got stuck with it. I have an alternative where you specify the file name for each reaction but it hasn't been finalized - if you'd prefer I can finish that up quickly and replace those as part of this PR.

lindsayad replied:
It depends how much work it is. It would be nice to not have files with spaces in the names for too long in the repo

keniley1 replied:
It shouldn't be bad. The groundwork is already in place. I should have it done by COB tomorrow.

keniley1 replied:
@lindsayad Alright, I couldn't quite get PetscJacobianTester working well. I tried scaling the mesh which helped a tiny bit, but when I tried scaling the time units I got a different answer than I did with unscaled timesteps, so I don't really trust that. However, it did converge and the relative difference was about an order of magnitude lower. I've added an example with a pretty loose relative tolerance 1e-4 just so it'll pass (without any timestep scaling), but that doesn't need to be final.

You're right, scaling everything to near unity is a little difficult with this system of equations. I'm not sure what the best path forward is.

lindsayad replied:
Ok, 1e-4 isn't terrible.

Am I a crazy person or does this file not still have spaces in the name?

keniley1 replied:
Nope, you're not crazy. I'm just a fool. I changed all the filenames in the tests directory but not in problems. Give me a minute...

keniley1 replied:
Sorry about that. Should be fixed now.

keniley1 commented:
@lindsayad No, I am not! I'll give it a shot.

lindsayad commented:
> @lindsayad No, I am not! I'll give it a shot.

You can find some examples throughout MOOSE; you can find one in `moose/test/tests/kernels/ad_coupled_value/tests`

keniley1 commented:
@lindsayad  Well, good news is that it was easy to include! Bad news is it fails. For timestep one I'm getting a ||J - Jfd||_F/||J||_F value of ~2e-5 at the first timestep and ~9e-6 at the 5th. That's not horrible, but since the tests suggests that a value of 1e-8 is considered "correct" I'm assuming something's wrong here. I'll take a look to see if I can find any issues. 

Anecdotally, AD runs very well. I'm interested to see where the error(s) could be.

lindsayad commented:
Keep in mind that that 1e-8 value is based on a "well-scaled" problem. If
you are doing everything AD, the chances are almost as good that the error
is in the finite-differenced Jacobian instead of the AD produced Jacobian.

If possible for Jacobian testing, you would want almost all quantities to
be close to unity (mesh size, time step size, material properties), but
that may difficult for you to construct.

On Thu, Mar 19, 2020 at 4:06 PM Shane Keniley <notifications@github.com>
wrote:

> *@keniley1* commented on this pull request.
> ------------------------------
>
> In problems/ad_argon_chemistry/reaction_em + Ar -> em + Ar*.txt
> <https://github.com/shannon-lab/zapdos/pull/54#discussion_r395368060>:
>
> > @@ -0,0 +1,200 @@
> +2.9270e-01 0.0000e+00
>
> It shouldn't be bad. The groundwork is already in place. I should have it
> done by COB tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/shannon-lab/zapdos/pull/54#discussion_r395368060>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4BYPR2DD5XONCAFPSTRIKQRFANCNFSM4LO2N4CQ>
> .
>

keniley1 commented:
I modified Crane to allow user-input filenames and changed all of the offensive files! @csdechant  once this PR goes through (or if you update crane on your own), you can use whatever file names you want. If you use the action, just put it in parenthesis. e.g.

em + Ar -> em + Ar*  : EEDF   [-11]   (filename)

keniley1 commented:
> Great work @keniley1. Have you been able to do any new stuff with AD that you weren't able to do before?

Not quite yet! This was just a stepping stone. The first thing I want to try is using mixture-averaged diffusion coefficients for heavy species. AD will make that almost trivial, and it will get this model pretty darn close to COMSOL's. I also wanted to see if this would help my 2D simulation. It was always very finnicky, particularly with that Neumann circuit BC which for some reason refused to run. I needed to use the Dirichlet version with PJFNK instead. Hopefully this will help!

(I'd also like to try a linear formulation with SUPG if I have time!)

csdechant commented:
@keniley1 Regarding SUPG, I tried making a plasma version of SUPG for Zapdos in the past (in branch: https://github.com/csdechant/zapdos/tree/WorkInProgress ). I never worked out all the bugs, but it's a start, if you wanted to look at it.

lindsayad commented:
> It was always very finnicky, particularly with that Neumann circuit BC which for some reason refused to run.

So AD doesn't run through postprocessors; postprocessors depend on all degrees of freedom and at this time we feel like it would be prohibitively expensive to store and operate with all those dofs. We may support it sometime in the future, but not currently. I forget whether that circuit thingy uses a postprocessor...I think one BC class does
